File: src/backend/app/schemas/agent.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.enums import AgentRunStatus, GenerationMode
from app.schemas.tools import GeneratedPlan, PlanExclusion, PlanItem, Risk, TimetableSlot


class AgentRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    run_id: str
    student_id: str
    generation_mode: GenerationMode
    provider: str | None
    status: AgentRunStatus
    latency_ms: int | None
    error: str | None
    created_at: datetime


class ToolCallRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    tool_call_id: str
    run_id: str
    tool_name: str
    input_json: Any
    output_json: Any
    duration_ms: int | None
    success: bool
    created_at: datetime


class ChatRequest(BaseModel):
    message: str = Field(min_length=1)
    student_id: str | None = None
    term_id: str | None = None


class ChatResponse(BaseModel):
    reply: str
    intent: Literal["plan", "course", "risk", "greet", "refuse"]
    generation_mode: GenerationMode
    used_template: bool
    run_id: str | None = None


class RecommendedCourseView(BaseModel):
    course_code: str
    course_name: str
    note: str


class PlanGenerateRequest(BaseModel):
    student_id: str
    term_id: str | None = None
    target_credit_load: int | None = None
    include_retakes: bool = True
    note: str | None = None


class ExplainedPlanResponse(BaseModel):
    student_id: str
    term_id: str
    generation_mode: GenerationMode
    explanation_source: Literal["llm", "template"]
    explanation: str
    items: list[PlanItem]
    exclusions: list[PlanExclusion]
    timetable: list[TimetableSlot]
    total_credits: int
    course_count: int
    warnings: list[str]
    risks: list[Risk]
    recommended: list[RecommendedCourseView]
    names: dict[str, str]
    run_id: str | None = None
    plan: GeneratedPlan

File: src/frontend/src/services/Agent.ts
import { apiRequest } from './api';
import { getCurrentStudentId } from './session';

export const DEFAULT_TERM_ID = 'TERM-2026-1';

export type AgentChatResponse = {
  reply: string;
  intent: string;
  generation_mode: 'LLM' | 'Fallback';
  used_template: boolean;
  run_id: string | null;
};

export type PlanItem = {
  course_id: string;
  course_code: string;
  section_id: string;
  credits: number;
  selection_reason: string;
};

export type PlanExclusion = {
  course_id: string;
  course_code: string;
  reason: string;
};

export type TimetableSlot = {
  section_id: string;
  course_code: string;
  meeting_type: string;
  day_of_week: string;
  start_time: string;
  end_time: string;
  room: string;
};

export type AgentRisk = {
  code: string;
  severity: string;
  message: string;
  course_codes: string[];
};

export type RecommendedCourseView = {
  course_code: string;
  course_name: string;
  note: string;
};

export type ExplainedPlanResponse = {
  student_id: string;
  term_id: string;
  generation_mode: 'LLM' | 'Fallback';
  explanation_source: 'llm' | 'template';
  explanation: string;
  items: PlanItem[];
  exclusions: PlanExclusion[];
  timetable: TimetableSlot[];
  total_credits: number;
  course_count: number;
  warnings: string[];
  risks: AgentRisk[];
  recommended: RecommendedCourseView[];
  names: Record<string, string>;
  run_id: string | null;
};

export function chatWithAgent(message: string, termId = DEFAULT_TERM_ID) {
  return apiRequest<AgentChatResponse>('/api/agent/chat', {
    method: 'POST',
    body: JSON.stringify({
      message,
      student_id: getCurrentStudentId(),
      term_id: termId,
    }),
  });
}

export function generateExplainedPlan(input: {
  targetCredits?: number;
  note?: string;
  termId?: string;
}) {
  return apiRequest<ExplainedPlanResponse>('/api/agent/plan', {
    method: 'POST',
    body: JSON.stringify({
      student_id: getCurrentStudentId(),
      term_id: input.termId ?? DEFAULT_TERM_ID,
      target_credit_load: input.targetCredits,
      note: input.note || null,
    }),
  });
}
